src/sparse_blas/backends/mkl_common/mkl_spmv.cxx

```diff
@@ -0,0 +1,279 @@
+    }
+
+    if ((A_view.type_view == oneapi::mkl::sparse::matrix_descr::symmetric ||
+         A_view.type_view == oneapi::mkl::sparse::matrix_descr::hermitian) &&
+        opA == oneapi::mkl::transpose::conjtrans) {
+        throw mkl::unimplemented(
+            "sparse_blas", function_name,
+            "The backend does not support Symmetric or Hermitian matrix with `conjtrans`.");
+    }
+}
+
+void spmv_buffer_size(sycl::queue &queue, oneapi::mkl::transpose opA, const void *alpha,
+                      oneapi::mkl::sparse::matrix_view A_view,
+                      oneapi::mkl::sparse::matrix_handle_t A_handle,
+                      oneapi::mkl::sparse::dense_vector_handle_t x_handle, const void *beta,
+                      oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+                      oneapi::mkl::sparse::spmv_alg /*alg*/,
+                      oneapi::mkl::sparse::spmv_descr_t spmv_descr, std::size_t &temp_buffer_size) {
+    // TODO: Add support for external workspace once the close-source oneMKL backend supports it.
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    bool is_beta_host_accessible = detail::is_ptr_accessible_on_host(queue, beta);
+    check_valid_spmv(__func__, opA, A_view, A_handle, x_handle, y_handle, is_alpha_host_accessible,
+                     is_beta_host_accessible);
+    temp_buffer_size = 0;
+    spmv_descr->buffer_size_called = true;
+}
+
+inline void common_spmv_optimize(sycl::queue &queue, oneapi::mkl::transpose opA, const void *alpha,
+                                 oneapi::mkl::sparse::matrix_view A_view,
+                                 oneapi::mkl::sparse::matrix_handle_t A_handle,
+                                 oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+                                 const void *beta,
+                                 oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+                                 oneapi::mkl::sparse::spmv_alg alg,
+                                 oneapi::mkl::sparse::spmv_descr_t spmv_descr) {
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    bool is_beta_host_accessible = detail::is_ptr_accessible_on_host(queue, beta);
+    check_valid_spmv(__func__, opA, A_view, A_handle, x_handle, y_handle, is_alpha_host_accessible,
+                     is_beta_host_accessible);
+    if (!spmv_descr->buffer_size_called) {
+        throw mkl::uninitialized(
+            "sparse_blas", __func__,
+            "spmv_buffer_size must be called with the same arguments before spmv_optimize.");
+    }
+    spmv_descr->optimized_called = true;
+    spmv_descr->last_optimized_opA = opA;
+    spmv_descr->last_optimized_A_view = A_view;
+    spmv_descr->last_optimized_A_handle = A_handle;
+    spmv_descr->last_optimized_x_handle = x_handle;
+    spmv_descr->last_optimized_y_handle = y_handle;
+    spmv_descr->last_optimized_alg = alg;
```

Comment:
These must be moved to `spmv_buffer_size` call. The temporary workspace size returned by the `spmv_buffer_size` API depends on all of these parameters. If any of them has changed since the `spmv_buffer_size` call for a given descriptor, the buffer size is no longer valid (may be lower, which is fine, but may also be higher, which is not fine.)

Reply:
Resolving as per my comment, https://github.com/oneapi-src/oneMKL/pull/500#discussion_r1739212240